### DISCUSSION TITLE
Exception Handling

### DISCUSSION THREAD

**Discussion by tsaxking** [mattpocock/ts-reset#149] [Ideas]
Hi I love this package! It works well and helps me write robust code. It's also very helpful because I'm a project manager for a team of students and it's very easy to write spaghetti code, this helps the end user manage that. I have a proposition though,

One thing that bothers me about Typescript that other languages such as Rust have is robust exception handling. During compile time, I'd like to have tsc tell me that I could have an uncaught exception using types. 

An example of the workflow:
```typescript
// what you provide, (obviously imported, but for the sake of clarity...)
enum Exception {
  fileNotExist,
  corrupted,
  noDirectory,
  noError
};

const isError = (e: Exception|any) => e !== Exception.noError;


// user workflow
const file: Exception|string = fs.readFileSync('file.txt', 'utf-8');

if (isError(file)) {
  switch (file) {} // handle each case
} else {
  // file works
}
```
So you would be introducing a new enum called `Exception` that changes values for each potential error. Like fs.readFileSync() enum would have the above errors that the end user could handle using a function you also provide. I understand this example wouldn't technically work considering fs.readFileSync would throw an error at runtime and thus stop the process, however something similar to this would be useful. You'd probably have to re-provide every function you add enums for, but there is probably a better way to handle this.

I don't think it'd work like the rest of the project because the end user would need to import the enums and that function, but it's food for thought.

**Comment by mattpocock** [comment]
Hey! This is a fine idea but definitely out of scope for this project.
